### DISCUSSION TITLE
How to get authorization decision with resource URI?

### DISCUSSION THREAD

**Discussion by VerdonTrigance** [keycloak/keycloak#10044] [Q&A] [answered]
I have two clients: public js front and confidential back (asp .net core) with UMA configured. I have some resources like /api/v1/survey/123 and others in my confidential resource server. My front client doesn't know anything about the way I authorize him (and it shouldn't). It just send request to specific url like /api/v1/survey/123. It means that my authorization handler in backend should check this request with this url. Again, client doesn't know which resource id associated with url '/api/v1/survey/123'. That's why I looking for a way I'd be able to check authority of this request by passing url to keycloak. Now I have to use PAT to get resource id from 'http://${host}:${port}/auth/realms/${realm_name}/authz/protection/resource_set?uri=/api/v1/survey/123' and after that send request to 'http://${host}:${port}/auth/realms/${realm}/protocol/openid-connect/token' with decision mode 'decision'. 

It works fine until I don't have resources with wildcard URI like this '/api/v1/survey/123/*'. If I use 'http://${host}:${port}/auth/realms/${realm_name}/authz/protection/resource_set?uri=/api/v1/survey/123/report' to find resource id it returns empty. I assume I shouldn't implement (duplicate) policy decision algorithm or parse (or regex) url on my side and it should be done by keycloak. But I don't see any other API to make it possible.

Maybe I do something wrong in my approach and you know better way?

**Comment by DanApostol** [comment]
I don't know if you still need a solution, but take a look on this link: https://stackoverflow.com/questions/58831162/fetch-resource-given-partial-url-path-or-based-on-a-regex-pattern-using-keycloak

It looks like setting matchingUri=true as query param does the magic of finding resources by pattern matching.

Good luck!

**Reply by VerdonTrigance** [reply]
I abandoned this project already, but I think wildcard in this link works in opposite way. They send wildcard TO keycloak and it returns answer based on list of uri. But what I need is to pass specific uri to keycloak and get response based on wildcard inside available resources.

**Comment by DanApostol** [comment, marked as answer]
En contraire, what I've tested was the following:

resource /test founded by calling resource_set?uri=/test
resource /test/random not found by calling resource_set?uri=/test/random, even it resource /test/* was defined in Authorization services in KC.

But:
if I'm calling resource_set?uri=/test/random&matchingUri=true, then resource is found (it is the resource defined with uri = /test/*)

So I guess that it should work as designed. Defining resources with wildcards in KC, and sending the full specific uri to resource_set endpoint, which will search for resources with wildcards which matches the specific uri, after trying to find the exact resource...

Thanks and regards.

**Reply by VerdonTrigance** [reply]
Thank you for your tests. I’m not in that project anymore, but I guess we might mark this as an answer. By the way can you show me where did you find matchingUri param documentation?

**Comment by jeisenbach-schulzst** [comment]
I know it's an old question but you can do it in one POST request like so:
```
curl -X POST \
  http://${host}:${port}/realms/${realm}/protocol/openid-connect/token \
  -H "Authorization: Bearer ${access_token}" \
  --data "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" \
  --data "response_mode=decision" \
  --data "audience={resource_server_client_id}" \
  --data "permission=/myResourceUri#read" \
  --data "permission_resource_format=uri"
  --data "permission_resource_matching_uri=true"
```
